### ISSUE TITLE
Extract WA password from device [Android users]

### ISSUE THREAD

**Issue by mgp25** [venomous0x/WhatsAPI#983]
![WAPW](http://cl.ly/image/070H3N2z2s2e/download/ic_launcher.png)

Hello guys, i have good news for the android users! Many of you asked me how to extract WhatsApp password from Android devices. @GitsUp has coded an .apk based on my code released here #895 and also you can find it in my [website](https://mgp25.com).

**Things i need to update (TODO list):**
- Extract identity ( I'm on the way people :p )
- Log useful http requests

**Useful information:**
I made a wiki in where i am explaining advanced stuff on reverse engineering WhatsApp and stuff (tokens, http requests, etc...) [WA Wiki!](http://wa.mgp25.com)

**How to contact me?**
Twitter: @_mgp25
Mail: me [ at ] mgp25.com

[**Download:**](https://mgp25.com/wa/wapw2.apk) wapw2.apk
**SHA1:** CA465128E4C4B4A4F70B37E968F62C302AA8A3E1
**MD5:** DB753933E47926DD131B097E069DB1D4

**Requirements**

You need your device to be rooted. If you don't know how to root your device, search in google :)

**How to use it?**

Easy! Follow this steps:

1. Open WA PW

![wapw](http://cl.ly/image/2G1p0h2K033b/SC20140925-005038.png)

2. Click on 'Extract pw!' and give root privileges to the apk

![root](http://cl.ly/image/2J1t1t121C1x/SC20140925-005047.png)

3. Now you have your WA PW!

![pw](http://cl.ly/image/020s0Y0V2i1Z/SC20140925-005055.png)

4. You can send it to your computer via WhatsApp, email...

![ta-chan](http://cl.ly/image/3o1V3A2k212e/SC20140925-005101.png)

**How it works?**

Is explained in #895 and also was explained in Yowsup repository.

**Me or pw file not found?**

You need to have an active number in your WhatsApp (Device)

**Additional info**

You can hook classes from WhatsApp to get data, like:

App$me.java

```java
package com.whatsapp;

import java.io.Serializable;

public class App$Me
  implements Serializable
{
  private static final long serialVersionUID = -1952119322404952662L;
  public String cc;
  public String jabber_id;
  public String number;

  App$Me(String paramString1, String paramString2)
  {
    this.cc = paramString1;
    this.number = paramString2;
    this.jabber_id = null;
  }
}
```

And [zw.java](http://cl.ly/code/3Z1Q050M1w1M)

This files as examples :p

**HEY! I'm an iPhone users!**

Don't worry! You can use [this utility](https://github.com/mgp25/WA-password-extractor) i made!

**Comment by shishirtanwar** [comment]
Doesn't install. Keeps saying 'Not Installed'

**Comment by shishirtanwar** [comment]
Okay, got it to install.....turns out if you have smorra's app installed it doesn't let wawp2 to get installed and vice-versa.
Anyway, on clicking extract button it says - "Can't access WA files. Are you root?"
But it is rooted and I already gave it SuperUser Access.
All other rooted apps are working fine.

**Comment by LukiLeu** [comment]
Hi 
I used your apk to extract the password. But when i try to use it in "Wassapp", it always mentions that my password is incorrect. Do I have to decode the retrieved password or something similar?
thanks

**Comment by waiferv** [comment]
I click on the Extract button but then I only get:
"Can't access WA files. Are you root?"

Device is rooted!!!

**Comment by mgp25** [comment]
@waiferv Follow guide from WhatsAPI-Official https://github.com/mgp25/WhatsAPI-Official/wiki/Extracting-password-from-device

Regards

**Comment by waiferv** [comment]
Still 

"Can't access WA files. Are you root?" 

with apk from that guide :(

**Comment by mgp25** [comment]
@waiferv give WhatsApp root privileges and then try again. Remember to use WhatsAPI Official repo, this one is inactive

**Comment by cooler-farmer** [comment]
Doesn't work android 5.0.2 for me :(

**Comment by w00000w5** [comment]
thanks mgp25

**Comment by mgp25** [comment]
@cooler-farmer @w00000w5 Use the apk from WhatsAPI Official repo: https://github.com/mgp25/WhatsAPI-Official

https://github.com/mgp25/WhatsAPI-Official/wiki/Extracting-password-from-device

That should work

Regards

**Comment by cooler-farmer** [comment]
@mgp25 it shows "couldn't read whatsapp files". 
I granted it root permissions :(
Using android 5.0.2

**Comment by vlnsnewton** [comment]
Dear All.

I have extracted whatsapp password but from where can i extract identity??
and can i use these whatsapp password and identity credentials in https://github.com/mgp25/WhatsAPI-Official


help me in this case in am using bluestack

**Comment by Killer1337** [comment]
on android 5.0.2 I also get "couldn't read whatsapp files"

Is there a solution to this?

**Comment by Killer1337** [comment]
here is a windows tool that does the same: https://github.com/shirioko/WART

**Comment by mgp25** [comment]
@Killer1337 it doesnt do the same, one thing is register the number, the other is extract the password. You should go here: https://github.com/mgp25/WhatsAPI-Official its more updated and have all the documentation about the API.

Regards

**Comment by jesussales** [comment]
@mgp25 if i want generate a pw file??? The extractor read a pw file and
make proccess to extract The pass but if i have a pass. .. do you know how
create The file? Or how edit pw to inject the pass??

**Comment by mgp25** [comment]
@jesussales Yes, its possible to generate the password but why? Use WhatsAPI Official instead of this repo.

**Comment by jesussales** [comment]
Only want to Know.... can you say me how create it?

**Comment by jharihar** [comment]
Hi i am trying to get the password from emulator, it is giving error as cant access wp data. are you root. 
please help me.. i have the pw file from abd. could u pls extract the password

**Comment by freesrc** [comment]
"Extract identity ( I'm on the way people :p )"

Do you mean that you are still struggling on getting wa identity?

**Comment by assegaf** [comment]
@mgp25 any instruction extracting pw wa on Nokia S40 device, its Java J2ME based apps.

**Comment by WarpigtheKiller** [comment]
Crashed instantly on 5.0.x.

**Comment by aghict** [comment]
Is there any way to get WhatsApp passsword without root access?

**Comment by marcelocecin** [comment]
INSTALL_FAILED_CPU_ABI_INCOMPATIBLE

**Comment by francwalter** [comment]
Obviously this version, and neither the newer version "Password Extractor (3)" doesn't work anymore.
On CM12.1 (jfltexx) I get with this and the newer apk, the same error as the others above:

> Couldn't read WhatsApp files. Are you
> root?

Yes, clear that I am root.
And I can get the "pw" file with adb though.
But I wont start to compile the tools because I am afraid this will result in the same.
Is there now an actual working way?

Thanks. frank

**Comment by catch2patch** [comment]
Same here, but getting password with bash works perfectly.

**Comment by francwalter** [comment]
What do you mean exactly with:

> getting the password with bash

do you mean compiling the tools?

**Comment by mgp25** [comment]
https://github.com/WHAnonymous/Chat-API/wiki/Extracting-password-from-device

From now on, use this repo: https://github.com/WHAnonymous/Chat-API

**Comment by francwalter** [comment]
Yes, but where can I ask about issues with that?
This is a wiki not an issue tracker :(
I read the [Extracting password from device](https://github.com/WHAnonymous/Chat-API/wiki/Extracting-password-from-device) and don't understand what is ment with this part (a bash-script):

```
#!/bin/bash

TELNUM='99123456789'
if ! [ -r pw ]; then exit 1; fi

dd if=pw of=pw_salt bs=1 skip=29 count=4
hexdump -e '2/1 "%02x"' pw_salt
dd if=pw of=pw_iv bs=1 skip=33 count=16
dd if=pw of=pw_ekey bs=1 skip=49 count=20

echo -n 'c2991ec29b1d0cc2b8c3b7556458c298c29203c28b45c2973e78c386c395' | xxd -r -p > pbkdf2_pass.bin
echo -n $TELNUM | hexdump -e '2/1 "%02x"' | xxd -r -p >> pbkdf2_pass.bin
```

It is not explained in the code, what this does or how it is used. I see that there is a variable TELNUM but I don' understand the "c2991ec29b1d0cc2b8c3b7556458c298c29203c28b45c2973e78c386c395" 
What is that? Do I need to change it? What do I put there?
Questions that I cannot post there.

EDiT: OK, I understand. I have put an [issue #1076](https://github.com/WHAnonymous/Chat-API/issues/1076) into that wiki about this.
I ran all these compiling and commands, but I doubt that it works with that unchanged hex. Will try anyway.
OK this works, nothing to change beside the TELNUM

**Comment by francwalter** [comment]
Thank you by the way for this great code!!!

**Comment by LavanyaLea** [comment]
@francwalter and @mgp25 I can't use the apk including the one I downloaded from https://github.com/WHAnonymous/Chat-API/wiki/Extracting-password-from-device.
it says "Couldn't read WhatsApp files. Are you root?" - same problem that other users have mentioned.
I am running on Bluestack on Mac.
Tried to use bash, however struggled straight away on this:
- Once the file has been copied to the SD card, use ADB again to copy it to your computer.

$ adb pull /sdcard/pw

it says pw is read only.

(bash looks terribly complicated to me who's inexperienced. Did anyone manage to get the apk working?)

thanks!
